CoreScriptsRoot/CoreScripts/PerformanceStatsManagerScript.lua

```diff
@@ -0,0 +1,219 @@
+    AddButton(displayType, i)
+  end
+end
+
+function OnButtonToggled(toggledDisplayType) 
+  local toggledButton = statsButtonsByType[toggledDisplayType]
+  local selectedState = toggledButton._isSelected
+  selectedState = not selectedState
+  
+  print("OnButtonToggled type=", toggledDisplayType, "  selected=", selectedState)
+  
+  if (selectedState) then 
+    currentDisplayType = toggledDisplayType
+  else
+    currentDisplayType = nil
+  end
+  
+  UpdateButtonSelectedStates()
+  UpdateViewerVisibility()
+end
+
+function UpdateButtonSelectedStates()
+  print("UpdateButtonSelectedStates ", currentDisplayType)
+  for i, buttonType in ipairs(StatsUtils.AllStatDisplayTypes) do
+      local button = statsButtonsByType[buttonType]
+      button:SetIsSelected(buttonType == currentDisplayType)
+  end  
+end
+
+function UpdateViewerVisibility()
+  -- If someone is on, show the Viewer.
+  -- FIXME(dbanks)
+  -- Configure with details of the dude currently selected.
+  print("UpdateViewerVisibility")
+  print("currentDisplayType = ", currentDisplayType)
+  
+  if (currentDisplayType == nil) then 
+    statsViewer:SetVisible(false)
+    statsViewer:SetStatsAggregator(nil)
+  else
+    statsViewer:SetStatsDisplayType(currentDisplayType)
+    local aggregatorType = StatsUtils.DisplayTypeToAggregatorType[currentDisplayType]
+    statsViewer:SetStatsAggregator(allStatsAggregators:GetAggregator(aggregatorType))
+    
+    statsViewer:SetVisible(true)
+  end
+end
+
+function AddButton(displayType, index) 
+  -- Configure size and position of button.
+  -- Configure callback behavior to toggle
+  --    button on or off and show/hide viewer.
+  -- Parent button in main screen.
+  local button = statsButtonsByType[displayType]
+  
+  button:SetGUIParent(masterFrame)
+  local aggregatorType = StatsUtils.DisplayTypeToAggregatorType[displayType]
+  button:SetStatsAggregator(
+    allStatsAggregators:GetAggregator(aggregatorType))
+  
+  local fraction = 1.0/StatsUtils.NumButtonTypes
+  local size = UDim2.new(fraction, 0, 0.1666, 0)
+  local position = UDim2.new(fraction * (index - 1), 0, 0, 0)
+  button:SetSizeAndPosition(size, position)
+  
+  button:SetToggleCallbackFunction(OnButtonToggled)
+end
+
+function ConfigureStatViewerInMasterFrame()
+  -- Set up the widget that shows currently selected button.
+  statsViewer:SetGUIParent(masterFrame)
+  
+  local size = UDim2.new(0.5, 0, 0.5, 0)
+  local position = UDim2.new(0.5, 0, 0.25, 0)
+  statsViewer:SetSizeAndPosition(size, position)
+end
+
+function UpdatePerformanceStatsVisibility() 
+  masterFrame.Visible = GameSettings.PerformanceStatsVisible
+end
+
+
+--[[ Top Level Code ]]--
+-- If flag is not enabled, bounce.
+if not showPerformanceStatsInGui then 
+	return
+end
+ 
+ 
+-- Set up our GUI.
+ConfigureMasterFrame()
+ConfigureStatButtonsInMasterFrame()
+ConfigureStatViewerInMasterFrame()
+
+-- Watch for changes in local player.
+PlayersService.PlayerAdded:connect(UpdateLocalPlayer)
```

Comment:
it might be possible, but in that case I'd still get rid of these two connections to UpdateLocalPlayer. Instead, I'd do this below, where we explicitly call UpdateLocalPlayer:

` localPlayer = PlayersService.LocalPlayer
  while not localPlayer do
     PlayersService.Changed:wait()
     localPlayer = PlayersService.LocalPlayer
  end`

This way we don't have to listen to extraneous events that won't ever change the thing you are interested in